### ISSUE TITLE
No gradient for bias in embedding layer

### ISSUE THREAD

**Issue by ArneNx** [rwth-i6/returnn#182]
When I add a bias to my embedding layer, I see the following warning:
```
Warning: No slot_var found for variable <tf.Variable 'source_embed_raw/b:0' shape=(512,) dtype=float32_ref>, slot_name 'm'. Maybe no gradient for this var?
Warning: No slot_var found for variable <tf.Variable 'source_embed_raw/b:0' shape=(512,) dtype=float32_ref>, slot_name 'v'. Maybe no gradient for this var?
```
The corresponding part of my config looks like this:
```python
'source_embed_raw': { 'activation': None,
                        'class': 'linear',
                        'forward_weights_init': "variance_scaling_initializer(mode='fan_in', distribution='uniform', scale=1.0)",
                        'from': ['src_mask'],
                        'n_out': 512,
                        'with_bias': True},
```
where `src_mask` is a sparse tensor.
I also tried to simulate the same effect by multiplying a constant 1 with another linear layer and adding the result to `source_embed_raw` (instead of using a bias).
This gave me the same warning (for the new linear layer) and I also could see that the corresponding parameters are not updated throughout the training.

I was able to see that the `slot_var`s are in fact not present in the optimizer (Adam in my case). 
However, I'm not sure why this is the case. Does anyone have an idea?

**Comment by albertz** [comment]
Well, the warning does not necessarily say that there is no gradient. Only no slot for Adam. I'm not exactly sure though why this could be.

I think you should also get a warning if there is no gradient to the variable itself (maybe check that, not sure right now). So if you don't see that, it is probably all fine.

Which TF version?

How do you use `source_embed_raw` elsewhere in the network? Maybe the bias cancels out somehow?

**Comment by ArneNx** [comment]
I just found my problem. 
Turns out I had a layer with a tf.while_loop above this embedding. 
Without noticing, I had backprop deactivated for this while_loop. Therefore there is no gradient to the entire embedding. 

Nice to see that this gives a warning.
